### PR TITLE
MM-10677 memory fix to s3 upload

### PR DESCRIPTION
**Pull request by zetaab**
#### Summary
Currently mattermost is using `-1` size option when using s3 file upload. One writefile function call causes 576MB memory consumption and that can easily lead to server crash. Lets say we have 10 users uploading files at the same time, it takes already 5GB of memory.

So this fix is first copying the whole stream to buffer, and with that way we can define how much memory we really need. However, this is not good solution if the files are larger than original 576MB, because then the buffer size is going to be problem. At least for us we have maxfilesize limit lower than 576MB, so that is not problem for us.

#### Ticket Link
https://github.com/mattermost/mattermost-server/issues/8838
https://github.com/mattermost/mattermost-server/issues/9078

#### Checklist
- [x] All new/modified APIs include changes to the drivers

### CODE REVIEW

File: utils/file_backend_s3.go [hunk 1]
```diff
@@ -182,7 +183,9 @@ func (b *S3FileBackend) WriteFile(fr io.Reader, path string) (int64, *model.AppE
 	}
 
 	options := s3PutOptions(b.encrypt, contentType)
-	written, err := s3Clnt.PutObject(b.bucket, path, fr, -1, options)
+	var buf bytes.Buffer
+	buf.ReadFrom(fr)
```
Comment (lieut-data) [inline utils/file_backend_s3.go]: Can you handle the error returned by `ReadFrom`?

**Comment by lieut-data** [comment]
@jasonblais, would we consider this for 5.3? It's been a pain point for customers running Mattermost in "lowish" memory situations, and should be benign. It's a win for customers with "larger" memory situations too.

**Comment by zetaab** [comment]
we could still make this better. I think that we could pass maxfilesize to s3 writefile function. If maxfilesize is less than 576MB we could use this strategy and if larger than 576MB we should use `-1`

but its bigger change and no idea is that even needed

**Comment by jasonblais** [comment]
> @jasonblais, would we consider this for 5.3?

@lieut-data What is the risk here, if any? Potentially affecting file downloads to S3 in an unexpected way?

File: utils/file_backend_s3.go [hunk 2]
```diff
@@ -182,7 +183,12 @@ func (b *S3FileBackend) WriteFile(fr io.Reader, path string) (int64, *model.AppE
 	}
 
 	options := s3PutOptions(b.encrypt, contentType)
-	written, err := s3Clnt.PutObject(b.bucket, path, fr, -1, options)
+	var buf bytes.Buffer
+	_, err = buf.ReadFrom(fr)
+	if err != nil {
+		return written, model.NewAppError("WriteFile", "api.file.write_file.s3.app_error", nil, err.Error(), http.StatusInternalServerError)
```
Comment (lieut-data) [inline utils/file_backend_s3.go]: `written` isn't defined until line 191: this is the cause for the failing build.

**Comment by lieut-data** [comment]
@jasonblais, barring something like "S3 file uploads are completely broken" (which is easy to verify), I believe we're using strictly less memory in all cases now, since we've replaced the fixed-size minio buffer with one matching the filesize in question. I can't think of any specific new failure scenarios.

**Comment by jasonblais** [comment]
@zetaab Is that a failure scenario that also existed before? Or is it something that would be new?

**Comment by zetaab** [comment]
the 576MB has been the minimum and maximum reserved memory for single upload (single function call). However, now the minimum and maximum is the size of uploaded file. So if you upload large files this is not good change (see https://github.com/mattermost/mattermost-server/pull/9373#issuecomment-419094308 how we could still improve)

**Comment by jasonblais** [comment]
Confirmed file uploads work on S3.

It sounds like it would be safe to merge this to v5.3. Just a quick question: for those who do try to upload a file larger than 576MB, does it fail and give an error message to the user? It won't crash the server or anything?

I was going to test it out, but realized I don't have such a large file. We do have a few hundred servers that allow file uploads larger than 576MB.

**Comment by zetaab** [comment]
@jasonblais imo it should not crash if there is enough RAM memory in server. It crashes only if it runs out of memory.

**Comment by zetaab** [comment]
@jasonblais no idea was this what you wanted, but now its rebased against 5.3

**Comment by jasonblais** [comment]
Perhaps you can help with the rebasing @lieut-data? I'm actually not sure how it's done.

**Comment by zetaab** [comment]
@jasonblais now better?

**Comment by jasonblais** [comment]
Excellent @zetaab 👍

File: utils/file_backend_s3.go [hunk 3]
```diff
@@ -182,7 +183,12 @@ func (b *S3FileBackend) WriteFile(fr io.Reader, path string) (int64, *model.AppE
 	}
 
 	options := s3PutOptions(b.encrypt, contentType)
-	written, err := s3Clnt.PutObject(b.bucket, path, fr, -1, options)
+	var buf bytes.Buffer
+	_, err = buf.ReadFrom(fr)
+	if err != nil {
+		return 0, model.NewAppError("WriteFile", "api.file.write_file.s3.app_error", nil, err.Error(), http.StatusInternalServerError)
+	}
+	written, err := s3Clnt.PutObject(b.bucket, path, bytes.NewReader(buf.Bytes()), int64(len(buf.Bytes())), options)
```
Comment (crspeller) [inline utils/file_backend_s3.go]: `bytes.Buffer` is a reader and `len(buf.Bytes()) == buf.Len()` so I think we can simplify this:
```go
written, err := s3Clnt.PutObject(b.bucket, path, buf, int64(buf.Len()), options)
```

**Comment by zetaab** [comment]
I could not find documentation which says that bytes.Buffer is a reader

**Comment by crspeller** [comment]
@zetaab Sorry. I made a mistake in my example there. Interfaces need to be pointers. Here is the correct code:
```go
written, err := s3Clnt.PutObject(b.bucket, path, &buf, int64(buf.Len()), options)
```
Here is where it implements reader: https://golang.org/pkg/bytes/#Buffer.Read

**Comment by zetaab** [comment]
@crspeller

**Comment by crspeller** [comment]
Great work @zetaab !

@Belphemur It is scheduled as a fix for RC3. Let us know if it solves your issue.

**Comment by cpanato** [comment]
thanks @zetaab for this cool PR

**Pull request merged**